### PR TITLE
Properly implement percent decoding on URLs

### PR DESCRIPTION
**Pull request by KiChjang**
This will fix a couple of WPTs that were failing before.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22620)

### CODE REVIEW

**Comment by SimonSapin** [comment]
This functionality already exists in the `url` and `percent-encoding` crates.

**Comment by KiChjang** [comment]
Ah ok, I'll pull in functions from those crates instead.

**Comment by jdm** [comment]
You may be replacing the : in the scheme now as well.

File: components/net/http_loader.rs [hunk 1]
```diff
@@ -412,17 +413,10 @@ fn obtain_response(
     // TODO(#21261) connect_start: set if a persistent connection is *not* used and the last non-redirected
     // fetch passes the timing allow check
     let connect_start = precise_time_ms();
-    // https://url.spec.whatwg.org/#percent-encoded-bytes
+    let url_bytes = url.clone().into_url().as_ref().bytes().collect::<Vec<u8>>();
```
Comment (nox) [inline components/net/http_loader.rs]:
```rust
let url_bytes = url.as_url().as_str().as_bytes();
```

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -412,17 +413,10 @@ fn obtain_response(
...
         .method(method)
-        .uri(
-            url.clone()
-                .into_url()
-                .as_ref()
-                .replace("|", "%7C")
-                .replace("{", "%7B")
-                .replace("}", "%7D"),
-        )
+        .uri(percent_decode(&url_bytes).decode_utf8_lossy().into_owned())
```
Comment (nox) [inline components/net/http_loader.rs]: Why the lossy decoding here? Hyper doesn't need a `str`, maybe use `PercentDecode::if_any`.

**Comment by KiChjang** [comment]
Ah, it's entirely possible that I'm percent decoding the username:password part of the URL!

**Comment by KiChjang** [comment]
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"application/xhtml+xml\" but got \"application/xhtml xml\"", 
    "stack": "xhr.onload<@http://web-platform.test:8000/_mozilla/mozilla/mime_sniffing_font_context.html:54:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1566:20\nTest.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1606:17\n", 
    "subtest": "XHR Content-Type has xhtml+xml", 
    "test": "/_mozilla/mozilla/mime_sniffing_font_context.html", 
    "line": 127458, 
    "action": "test_result", 
    "expected": "PASS"
}
```

This failure is very peculiar... is the percent decoding not working as intended?

**Comment by nox** [comment]
What exactly are you implementing here? There should be no percent-decoding bug in that crate, AFAIK. What you probably encountered is step 3.4. of https://url.spec.whatwg.org/#concept-urlencoded-parser

**Comment by SimonSapin** [comment]
I agree with nox: I don’t understand the premise of this PR. Why should we be doing percent-*decoding* just before making an HTTP request? The request’s path is percent-*encoded* on the wire. Also `Url::set_path` is based on (part of) the same parsing algorithm as `Url::parse`, which percent-*encodes* on the fly. So the `percent_decode_path_and_fragment` probably doesn’t actually do what its name suggest.

More generally, converting correctly from `url::Url` to `hyper::Request` should ideally not require such complicated logic. If a conversion is indeed needed, either or both libraries should be extended to make it easier for users.

**Comment by KiChjang** [comment]
So, for this particular test in `tests/wpt/web-platform-tests/eventsource/format-mime-trailing-semicolon.htm`, you can see the URL contains a `%3B` at the end, which should be decoded somewhere during the processing of the request, but wasn't since the test was failing.

In any case, the `replace` calls for `|`, `{` and `}` should call for a more generalized fix for percent-decoding on the URL string. If it is as you both say that we shouldn't be percent-decoding the URL at all, then I'm not sure whether our implementation is incorrect, or if the tests are wrong.

**Comment by nox** [comment]
```
± curl -iv http://w3c-test.org/eventsource/resources/message.py?mime=text/event-stream%3B
*   Trying 128.30.54.138...
* TCP_NODELAY set
* Connected to w3c-test.org (128.30.54.138) port 80 (#0)
> GET /eventsource/resources/message.py?mime=text/event-stream%3B HTTP/1.1
> Host: w3c-test.org
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Content-Type: text/event-stream;
Content-Type: text/event-stream;
< Server: BaseHTTP/0.3 Python/2.7.13
Server: BaseHTTP/0.3 Python/2.7.13
< Date: Fri, 11 Jan 2019 09:35:07 GMT
Date: Fri, 11 Jan 2019 09:35:07 GMT
< Content-Length: 13
Content-Length: 13

<
data: data
```

The issue isn't with the URL, but with the `Content-Type` header we end up parsing in `EventSource`'s constructor. I'm trying to follow the specs to see where the hell is this header supposed to be percent-decoded, but I'm failing at that for now.

**Comment by nox** [comment]
```
± nc w3c-test.org 80
GET /eventsource/resources/message.py?mime=text/event-stream%3B HTTP/1.1
Host: w3c-test.org

HTTP/1.1 200 OK
Content-Type: text/event-stream;
Server: BaseHTTP/0.3 Python/2.7.13
Date: Fri, 11 Jan 2019 09:41:58 GMT
Content-Length: 13

data: data
```

Just to be sure it's not cURL doing the decoding.

**Comment by nox** [comment]
Found the problem, `mime` considers `text/event-source;` and `text/event-source` to be different.

**Comment by nox** [comment]
https://github.com/hyperium/mime/issues/101

**Comment by nox** [comment]
Whether there is a bug upstream or not doesn't matter in our case though, given we should be actively ignoring parameters anyway.

**Comment by nox** [comment]
I made a patch that compares just the type and subtype to `(mime::TEXT, mime::EVENT_STREAM)` and discovered https://github.com/hyperium/mime/issues/102.

**Comment by nox** [comment]
Superseded by #22678.

**Pull request closed without merging**